### PR TITLE
[feat]: [CCM-9630]: API support for forecasting monthly budget amounts

### PR DESCRIPTION
**Pull request by iamanmolm**
## Describe your changes
Here to forecast cost on monthly basis of yearly period budget we are using simple regression method to achieve that.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/39373)

### CODE REVIEW

**Comment by iamanmolm** [comment]
trigger compile, ti0, ft1, ft2, ti1, ti2, ti3, ti4

**Comment by iamanmolm** [comment]
trigger ti0

File: 374-ceng-graphql/src/main/java/io/harness/ccm/graphql/core/budget/BudgetCostServiceImpl.java [hunk 1]
```diff
@@ -123,6 +130,112 @@ public double getLastPeriodCost(String accountId, String perspectiveId, long sta
...
+    List<QLCEViewGroupBy> groupBy = new ArrayList<>();
+    groupBy.add(
+        QLCEViewGroupBy.builder()
+            .timeTruncGroupBy(
+                QLCEViewTimeTruncGroupBy.builder().resolution(getTimeResolutionForBudget(BudgetPeriod.MONTHLY)).build())
+            .build());
+    long endTime = BudgetUtils.getEndTimeForBudget(startOfPeriod, period) - BudgetUtils.ONE_DAY_MILLIS;
+    List<QLCEViewFilterWrapper> filters = new ArrayList<>();
+    filters.add(viewsQueryHelper.getViewMetadataFilter(perspectiveId));
+    filters.add(viewsQueryHelper.getPerspectiveTimeFilter(startOfPeriod, AFTER));
```
Comment (iamanmolm) [inline 374-ceng-graphql/src/main/java/io/harness/ccm/graphql/core/budget/BudgetCostServiceImpl.java:182]: startOfPeriod should be startOfMonth.

File: 374-ceng-graphql/src/main/java/io/harness/ccm/graphql/core/budget/BudgetCostServiceImpl.java [hunk 2]
```diff
@@ -123,6 +130,112 @@ public double getLastPeriodCost(String accountId, String perspectiveId, long sta
...
+      String accountId, String perspectiveId, long startOfPeriod, BudgetPeriod period) {
+    Double[] result;
+    String cloudProviderTableName = bigQueryHelper.getCloudProviderTableName(accountId, UNIFIED_TABLE);
+    List<QLCEViewGroupBy> groupBy = new ArrayList<>();
+    groupBy.add(
+        QLCEViewGroupBy.builder()
+            .timeTruncGroupBy(
+                QLCEViewTimeTruncGroupBy.builder().resolution(getTimeResolutionForBudget(BudgetPeriod.MONTHLY)).build())
+            .build());
+    long endTime = BudgetUtils.getEndTimeForBudget(startOfPeriod, period) - BudgetUtils.ONE_DAY_MILLIS;
```
Comment (iamanmolm) [inline 374-ceng-graphql/src/main/java/io/harness/ccm/graphql/core/budget/BudgetCostServiceImpl.java:179]: endTime would be computed based on startOfPeriod.

File: 374-ceng-graphql/src/main/java/io/harness/ccm/graphql/core/budget/BudgetCostServiceImpl.java [hunk 3]
```diff
@@ -123,6 +130,112 @@ public double getLastPeriodCost(String accountId, String perspectiveId, long sta
...
+    SimpleRegression regression = new SimpleRegression();
+
+    // Getting last year actual cot data and feeding it to our model
+    Double[] lastYearCost = getLastYearMonthlyCost(accountId, perspectiveId, startOfPeriod, period);
+    for (int i = 0; i < BudgetUtils.MONTHS; i++) {
+      regression.addData(-(BudgetUtils.MONTHS - i), lastYearCost[i]);
+    }
+
+    // In case of future budgets where startOfPeriod is set in some time future
+    long forecastTime =
```
Comment (iamanmolm) [inline 374-ceng-graphql/src/main/java/io/harness/ccm/graphql/core/budget/BudgetCostServiceImpl.java:208]: startOfMonth(currentMonth)

File: 374-ceng-graphql/src/main/java/io/harness/ccm/graphql/core/budget/BudgetCostServiceImpl.java [hunk 4]
```diff
@@ -123,6 +130,112 @@ public double getLastPeriodCost(String accountId, String perspectiveId, long sta
...
+        BudgetUtils.getStartOfCurrentDay() < startOfPeriod ? startOfPeriod : BudgetUtils.getStartOfCurrentDay();
+
+    // Find months difference between startOfPeriod & forecastTime
+    int monthDiff = monthsBetween(new DateTime(startOfPeriod), new DateTime(forecastTime)).getMonths();
+
+    // Filling result with actual cost we have so far and feeding it to our model
+    Double[] result = getActualMonthlyCost(accountId, perspectiveId, startOfPeriod, period);
+    for (int i = 0; i <= monthDiff && monthDiff < BudgetUtils.MONTHS; i++) {
```
Comment (iamanmolm) [inline 374-ceng-graphql/src/main/java/io/harness/ccm/graphql/core/budget/BudgetCostServiceImpl.java:216]: i < monthDiff

Comment (iamanmolm) [reply 374-ceng-graphql/src/main/java/io/harness/ccm/graphql/core/budget/BudgetCostServiceImpl.java:216]: We will be forecasting for whole current month as well, even though we have let say 14 days of actual cost.

File: 490-ce-commons/src/main/java/io/harness/ccm/commons/entities/billing/Budget.java [hunk 5]
```diff
@@ -44,12 +45,17 @@ public final class Budget implements PersistentEntity, UuidAware, AccountAccess,
   @Size(min = 1, max = 80, message = "for budget name must be between 1 and 80 characters long") @NotBlank String name;
   BudgetScope scope; // referred to as "Applies to" in the UI
   BudgetType type;
+  BudgetBreakdown breakdown;
   Double budgetAmount;
+  Double[] budgetMonthlyCost;
```
Comment (iamanmolm) [inline 490-ce-commons/src/main/java/io/harness/ccm/commons/entities/billing/Budget.java:50]: Maybe we can introduce BudgetMonthlyBreakdown and club all these into this class.

File: 374-ceng-graphql/src/main/java/io/harness/ccm/graphql/core/budget/BudgetCostServiceImpl.java (hunk 2 above)
Comment (iamanmolm) [reply 374-ceng-graphql/src/main/java/io/harness/ccm/graphql/core/budget/BudgetCostServiceImpl.java:203]: Done

File: 374-ceng-graphql/src/main/java/io/harness/ccm/graphql/core/budget/BudgetCostServiceImpl.java (hunk 3 above)
Comment (iamanmolm) [reply 374-ceng-graphql/src/main/java/io/harness/ccm/graphql/core/budget/BudgetCostServiceImpl.java:232]: Done

File: 374-ceng-graphql/src/main/java/io/harness/ccm/graphql/core/budget/BudgetCostServiceImpl.java (hunk 4 above)
Comment (iamanmolm) [reply 374-ceng-graphql/src/main/java/io/harness/ccm/graphql/core/budget/BudgetCostServiceImpl.java:240]: Done

File: 374-ceng-graphql/src/main/java/io/harness/ccm/graphql/core/budget/BudgetCostServiceImpl.java (hunk 1 above)
Comment (iamanmolm) [reply 374-ceng-graphql/src/main/java/io/harness/ccm/graphql/core/budget/BudgetCostServiceImpl.java:206]: Done

File: 490-ce-commons/src/main/java/io/harness/ccm/commons/entities/billing/Budget.java (hunk 5 above)
Comment (iamanmolm) [reply 490-ce-commons/src/main/java/io/harness/ccm/commons/entities/billing/Budget.java]: Done

**Comment by iamanmolm** [comment]
trigger checkstyle, ti0, ti1, ti2, ti3, ti4

**Review by svmav — APPROVED** [review]
> Reviewed 3 of 11 files at r3, 11 of 11 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/39373)* status: :shipit: complete! ✓

**Pull request merged**